deepchem/data/dft_data.py

```diff
@@ -0,0 +1,375 @@
+"""
+Density Functional Theory Data
+Derived from: https://github.com/mfkasim1/xcnn/blob/f2cb9777da2961ac553f256ecdcca3e314a538ca/xcdnn2/entry.py
+"""
+from __future__ import annotations
+from abc import abstractmethod, abstractproperty
+from typing import List, Dict, Optional, Union
+import numpy as np
+try:
+    import dqc
+    from dqc.system.mol import Mol
+    from dqc.system.base_system import BaseSystem
+    from dqc.grid.base_grid import BaseGrid
+    from deepchem.utils.dftutils import KSCalc
+except ModuleNotFoundError:
+    raise ModuleNotFoundError("This data class requires dqc")
+
+
+class DFTSystem():
+    """
+    The DFTSystem class creates and returns the various systems in an entry object as dictionaries.
+
+    Examples
+    --------
+    >>> from deepchem.data.dft_data import DFTSystem
+    >>> systems = {'systems': [{
+        'type': 'mol',
+        'kwargs': {
+            'moldesc': 'Li 1.5070 0 0; H -1.5070 0 0',
+            'basis': '6-311++G(3df,3pd)'
+        }
+    }, {
+        'type': 'mol',
+        'kwargs': {
+            'moldesc': 'Li 0 0 0',
+            'basis': '6-311++G(3df,3pd)',
+            'spin': 1
+        }
+    }, {
+        'type': 'mol',
+        'kwargs': {
+            'moldesc': 'H 0 0 0',
+            'basis': '6-311++G(3df,3pd)',
+            'spin': 1
+        }
+    }]}
+    >>> output = DFTSystem.create(systems)
+
+    Returns
+    -------
+    List of dictionaries for all the different atoms/ions/molecules in an entry object.
+
+    References
+    ----------
+    Kasim, Muhammad F., and Sam M. Vinko. "Learning the exchange-correlation functional from nature with fully differentiable density functional theory." Physical Review Letters 127.12 (2021): 126403.
+
+    https://github.com/diffqc/dqc/blob/0fe821fc92cb3457fb14f6dff0c223641c514ddb/dqc/system/base_system.py
+    """
+
+    @classmethod
+    def create(self, system: Dict) -> DFTSystem:
+        """
+        Parameters
+        ----------
+        system: Dict
+            system is a dictionary containing information on the atomic positions,
+            atomic numbers, and basis set used for the DFT calculation.
+
+        Returns
+        -------
+        Creates and returns the system if it has not been created
+        manually during training step. Otherwise, return the previously created system.
+        """
+        return DFTSystem(system)
+
+    def __init__(self, system: Dict):
+        self.system = system
+
+    def get_dqc_mol(self, pos_reqgrad: bool = False) -> BaseSystem:
+        """
+        This method converts the system dictionary to a DQC system and returns it.
+        Parameters
+        ----------
+        pos_reqgrad: bool
+            decides if the atomic position require gradient calculation.
+        """
+        systype = self.system["type"]
+        if systype == "mol":
+            atomzs, atomposs = dqc.parse_moldesc(
+                self.system["kwargs"]["moldesc"])
```

Comment:
This has a lot of potential for bugs. Kwargs for system isn't documented anywhere. Better to make all fields explicit.